core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -657,17 +656,17 @@ public void clearInferred(Resource... contexts) throws SailException {
 		verifyIsOpen();
 		verifyIsActive();
 		synchronized (datasets) {
-			if (inferredSink == null) {
+			if (inferredOnlySink == null) {
 				IsolationLevel level = getIsolationLevel();
-				SailSource branch = branch(true);
-				inferredDataset = branch.dataset(level);
-				inferredSink = branch.sink(level);
-				explicitOnlyDataset = branch(false).dataset(level);
+				SailSource branch = branch(IncludeInferred.inferredOnly);
+				inferredOnlyDataset = branch.dataset(level);
+				inferredOnlySink = branch.sink(level);
+				explicitOnlyDataset = branch(IncludeInferred.explicitOnly).dataset(level);
 			}
 			if (this.hasConnectionListeners()) {
-				remove(null, null, null, inferredDataset, inferredSink, contexts);
+				remove(null, null, null, inferredOnlyDataset, inferredOnlySink, contexts);
 			}
-			inferredSink.clear(contexts);
+			inferredOnlySink.clear(contexts);
```

Comment:
@jeenbroekstra 

I found out that the SailSourceConnection will get all statements and remove them from the inferred sink, instead of getting just the inferred statements. So I fixed that by introducing a tri-state enum, essentially you can now specify: all, inferredOnly or explicitOnly. This is only for this class....and `branch(...)` is a private method, so no breaking changes. 

I don't have any decent benchmarks for the rdfs sail that add and remove statements, but I do have some for the SPIN sail and that's seeing a 20% speed bump for some of the benchmarks.

Reply:
I'm lost. Is this related to #1795 ?

Reply:
Not related no. This just makes clearing inferred statements faster.